tough/src/editor/mod.rs

```diff
@@ -223,30 +349,139 @@ impl RepositoryEditor {
+        }
+
+        let targets = self
+            .targets_struct
+            .as_mut()
+            .ok_or_else(|| error::Error::NoTargets)?;
+        // Find the parent targets for the role we are creating
+        let mut parent = if from == "targets" {
+            targets
+        } else {
+            targets
+                .targets_by_name(from)
+                .context(error::DelegateMissing { name: from })?
+        };
+        // Get the keys used to sign the new role
+        let (keyids, key_pairs) = if let Some(key) = key_source {
+            let mut keyids = Vec::new();
+            let mut key_pairs = HashMap::new();
+            for source in key {
+                let key_pair = source
+                    .as_sign()
+                    .context(error::KeyPairFromKeySource)?
+                    .tuf_key();
+                let key_id = RepositoryEditor::add_key(&mut parent, key_pair.clone())?;
+                key_pairs.insert(key_id.clone(), key_pair);
+                keyids.push(key_id);
+            }
+            (keyids, key_pairs)
+        } else {
+            // If we weren't given a new key source create a role using parent keys
+            let mut keys = Vec::new();
+            for key in parent
+                .delegations
+                .as_ref()
+                .ok_or_else(|| error::Error::NoDelegations)?
+                .keys
+                .keys()
+            {
+                keys.push(key.clone());
+            }
+
+            (keys, HashMap::new())
+        };
+        let delegations = parent
+            .delegations
+            .as_mut()
+            .ok_or_else(|| error::Error::NoDelegations)?;
+        // Create new targets for `role`
+        let mut new_targets = Targets::new(SPEC_VERSION.to_string(), version, expiration);
+        let mut new_delegations = Delegations::new();
+        new_delegations.keys.extend(key_pairs);
+        // Create a new delegations for `role`
+        new_targets.delegations = Some(new_delegations);
+        let threshold = NonZeroU64::new(keyids.len().try_into().context(error::InvalidInto {})?)
+            .context(error::InvalidThreshold {})?;
+        delegations.roles.push(DelegatedRole {
+            name,
+            keyids,
+            threshold,
+            paths,
+            terminating: false,
+            targets: Some(Signed {
+                signed: new_targets,
+                signatures: Vec::new(),
+            }),
+        });
+
         Ok(self)
     }
 
-    /// Remove all targets from this repo
-    pub fn clear_targets(&mut self) -> &mut Self {
-        self.existing_targets
-            .get_or_insert_with(HashMap::new)
-            .clear();
-        self.new_targets.get_or_insert_with(HashMap::new).clear();
-        self
+    /// Adds a key to the targets delegation role if not already present, and returns a result with the key id.
+    fn add_key(targets: &mut Targets, key: Key) -> Result<Decoded<Hex>> {
```

Comment:
It seems like this method is specific to adding a key for delegations? If so, let's name it as such.

Also, do keys need to be added to a specific delegated role? Is there a missing parameter to specify the correct role?